ocp_resources/service_account.py

```diff
@@ -1,9 +1,52 @@
+import kubernetes
 from ocp_resources.resource import NamespacedResource
 
 
 class ServiceAccount(NamespacedResource):
     """
-    Service Account object
+    https://docs.openshift.com/container-platform/4.15/rest_api/security_apis/serviceaccount-v1.html
     """
 
     api_version = NamespacedResource.ApiVersion.V1
+
+    def __init__(
+        self,
+        automount_service_account_token=None,
+        image_pull_secrets=None,
+        secrets=None,
+        **kwargs,
+    ):
+        """
+        Args:
+            automountServiceAccountToken (bool): indicates whether pods running as this service account should have an
+                                                 API token automatically mounted
+            imagePullSecrets (list): list of references to secrets in the same namespace to use for pulling pod images
+            secrets (list): list of secrets for the pods to use
+        """
+        super().__init__(**kwargs)
+        self.automount_service_account_token = automount_service_account_token
+        self.image_pull_secrets = image_pull_secrets
+        self.secrets = secrets
+
+    def to_dict(self):
+        super().to_dict()
+        if not self.yaml_file:
+            if self.automount_service_account_token:
+                self.res["automountServiceAccountToken"] = self.automount_service_account_token
+            if self.image_pull_secrets:
+                self.res["imagePullSecrets"] = self.image_pull_secrets
+            if self.secrets:
+                self.res["secrets"] = self.secrets
+
+    def create_service_account_token(self, expiration_seconds=None, audiences=None, bound_object_ref=None):
```

Comment:
please add docstring
Can each of the args be None and still passed?

based on https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/V1TokenRequestSpec.md, `audiences` is mandatory

Reply:
It uses defaults. ```>>> out = sa.create_service_account_token(expiration_seconds=600)
>>> print(out)
{'api_version': 'authentication.k8s.io/v1', 'kind': 'TokenRequest', 'metadata': {'annotations': None, 'creation_timestamp': datetime.datetime(2024, 4, 15, 22, 21, 25, tzinfo=tzutc()), 'deletion_grace_period_seconds': None, 'deletion_timestamp': None, 'finalizers': None, 'generate_name': None, 'generation': None, 'labels': None, 'managed_fields': [{'api_version': 'authentication.k8s.io/v1', 'fields_type': 'FieldsV1', 'fields_v1': {'f:spec': {'f:expirationSeconds': {}}}, 'manager': 'OpenAPI-Generator', 'operation': 'Update', 'subresource': 'token', 'time': datetime.datetime(2024, 4, 15, 22, 21, 25, tzinfo=tzutc())}], 'name': 'prometheus-k8s', 'namespace': 'openshift-monitoring', 'owner_references': None, 'resource_version': None, 'self_link': None, 'uid': None}, 'spec': {'audiences': ['https://kubernetes.default.svc'], 'bound_object_ref': None, 'expiration_seconds': 600}, 'status': {'expiration_timestamp': datetime.datetime(2024, 4, 15, 22, 31, 25, tzinfo=tzutc()), 'token': '<token_string>'}}
```

Reply:
This is not optional.

https://kubernetes.io/docs/reference/kubernetes-api/authentication-resources/token-request-v1/

Reply:
This is using https://docs.openshift.com/container-platform/4.15/authentication/bound-service-account-tokens.html, with the kubernetes default 'https://kubernetes.default.svc'. This is the reason, create_service_account_token() should not make audiences as a required param.

Reply:
As discussed not using TokenRequest resource, as it is not a api-resource.